AX: lineTextMarkerRangeForTextMarker fails on first character of line in Google Docs

minorninth opened the pull request:
#### 69bf3ef0cfdc8cf5748d335c0652a704093630a8
<pre>
AX: lineTextMarkerRangeForTextMarker fails on first character of line in Google Docs
<a href="https://bugs.webkit.org/show_bug.cgi?id=275870">https://bugs.webkit.org/show_bug.cgi?id=275870</a>
<a href="https://rdar.apple.com/130529132">rdar://130529132</a>

Reviewed by Tyler Wilcock.

nextLineEndPosition has some workarounds for cases where the standard VisiblePosition
endOfLine function returns null. When computing AccessibilityObject::lineRangeForPosition
we should use that so that we&apos;re more likely to get a valid line.

* LayoutTests/accessibility/mac/line-text-marker-range-for-text-marker-expected.txt: Added.
* LayoutTests/accessibility/mac/line-text-marker-range-for-text-marker.html: Added.
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::lineRangeForPosition const):

Canonical link: <a href="https://commits.webkit.org/280426@main">https://commits.webkit.org/280426@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/920ef6d325dc1d51cf1a710996905b873c29a996

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56598 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35925 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9071 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60212 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7039 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/58725 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43548 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7230 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/45846 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4925 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58628 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33773 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48845 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26707 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30554 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6177 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6042 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/52511 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6448 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/61895 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/510 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6554 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53105 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/512 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48904 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53098 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12524 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/437 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31753 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32839 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33923 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32585 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on LayoutTests/accessibility/mac/line-text-marker-range-for-text-marker.html:
```diff
@@ -0,0 +1,30 @@
+  let textControl = accessibilityController.accessibleElementById("textcontrol");
+  let fullRange = textControl.textMarkerRangeForElement(textControl.childAtIndex(0));
+  let marker = textControl.startTextMarkerForTextMarkerRange(fullRange);
+  let range = textControl.lineTextMarkerRangeForTextMarker(marker);
```
I think these could all be `const`, but doesn't matter too much.

twilco commented on LayoutTests/accessibility/mac/line-text-marker-range-for-text-marker.html:
```diff
@@ -0,0 +1,30 @@
+  let marker = textControl.startTextMarkerForTextMarkerRange(fullRange);
+  let range = textControl.lineTextMarkerRangeForTextMarker(marker);
+  window.rangeStr = (textControl.stringForTextMarkerRange(range) + "").trim();
+  output += expect("rangeStr", '"First line of text"');
```
I think we tend towards using `var rangeStr` for things that need to be used in `expect` rather than assigning the variable to `window`. I think it's a bit more resilient in case `window` eventually gains a property with that name (though obviously probably not going to happen for `rangeStr`)

twilco replied:
Though maybe a `var` would conflict in that case too 🤔 I'm fine leaving this as-is or using `var`

twilco commented on LayoutTests/accessibility/mac/line-text-marker-range-for-text-marker.html:
```diff
@@ -0,0 +1,30 @@
+<script src="../../resources/accessibility-helper.js"></script>
+<script src="../../resources/js-test.js"></script>
+</head>
+<body id="body">
```
Is this `id` necessary?

twilco reviewed (approved):
Nice!

minorninth committed:
AX: lineTextMarkerRangeForTextMarker fails on first character of line in Google Docs
https://bugs.webkit.org/show_bug.cgi?id=275870
rdar://130529132

Reviewed by Tyler Wilcock.

nextLineEndPosition has some workarounds for cases where the standard VisiblePosition
endOfLine function returns null. When computing AccessibilityObject::lineRangeForPosition
we should use that so that we're more likely to get a valid line.

* LayoutTests/accessibility/mac/line-text-marker-range-for-text-marker-expected.txt: Added.
* LayoutTests/accessibility/mac/line-text-marker-range-for-text-marker.html: Added.
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::lineRangeForPosition const):

Canonical link: https://commits.webkit.org/280426@main
